spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/pubsub/brave/PubSubTracing.java

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.trace.pubsub.brave;
+
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import brave.Span;
+import brave.SpanCustomizer;
+import brave.Tracer;
+import brave.Tracing;
+import brave.messaging.MessagingRequest;
+import brave.messaging.MessagingTracing;
+import brave.propagation.Propagation;
+import brave.propagation.Propagation.Getter;
+import brave.propagation.TraceContext.Extractor;
+import brave.propagation.TraceContext.Injector;
+import brave.propagation.TraceContextOrSamplingFlags;
+import brave.sampler.SamplerFunction;
+import com.google.cloud.pubsub.v1.MessageReceiver;
+import com.google.cloud.pubsub.v1.PublisherInterface;
+import com.google.cloud.pubsub.v1.stub.SubscriberStub;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PubsubMessage.Builder;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+
+/** Use this class to decorate your Pub/Sub publisher / subscriber and enable Tracing. */
+public final class PubSubTracing {
+	/** Used for local message processors in {@link PubSubTracing#nextSpan(PubsubMessage.Builder)}. */
+	static final Getter<PubsubMessage.Builder, String> GETTER = new Getter<PubsubMessage.Builder, String>() {
+		@Override public String get(PubsubMessage.Builder request, String key) {
+			return request.getAttributesOrDefault(key, null);
+		}
+
+		@Override public String toString() {
+			return "PubsubMessage.Builder::getAttributesOrThrow";
+		}
+	};
+
+	final Tracing tracing;
+
+	final Tracer tracer;
+
+	final Extractor<PubSubProducerRequest> producerExtractor;
+
+	final Extractor<PubSubConsumerRequest> consumerExtractor;
+
+	final Extractor<PubsubMessage.Builder> processorExtractor;
+
+	final Injector<PubSubProducerRequest> producerInjector;
+
+	final Injector<PubSubConsumerRequest> consumerInjector;
+
+	final Set<String> traceIdHeaders;
+
+	final TraceContextOrSamplingFlags emptyExtraction;
+
+	final SamplerFunction<MessagingRequest> producerSampler;
+
+	final SamplerFunction<MessagingRequest> consumerSampler;
+
+	final String remoteServiceName;
+
+	final boolean singleRootSpanOnReceiveBatch;
```

Comment:
How about `groupReceivedMessageSpans` or `groupPulledMessageSpans` or something? I was confused by the purpose at first.

Reply:
I was just trying to stay consistent with `KafkaTracing` and others in Brave.